libraries/lib-module-manager/AsyncPluginValidator.cpp

```diff
@@ -0,0 +1,237 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file AsyncPluginValidation.cpp
+
+  @author Vitaly Sverchinsky
+
+  Part of lib-module-manager library
+
+**********************************************************************/
+
+#include "AsyncPluginValidator.h"
+
+#include <optional>
+#include <mutex>
+
+#include "BasicUI.h"
+#include "IPCChannel.h"
+#include "IPCServer.h"
+#include "spinlock.h"
+#include "XMLTagHandler.h"
+
+#include "PluginIPCUtils.h"
+#include "PluginDescriptor.h"
+#include "PluginHost.h"
+#include "XMLFileReader.h"
+
+AsyncPluginValidator::Delegate::~Delegate() = default;
+
+class AsyncPluginValidator::Impl final : 
+   public IPCChannelStatusCallback,
+   public std::enable_shared_from_this<Impl>
+{
+   //Variables below are accessed from any thread
+
+   IPCChannel* mChannel{nullptr};
+   std::optional<wxString> mRequest;
+
+   //The main reason to use spinlock instead of std::mutex here
+   //is that in most cases there will be no attempts for simultaneous
+   //data access (except OnConnect/Disconnect cases), because of
+   //protocol design
+   spinlock mSync;
+
+   //Variable below is accessed only from the main/UI thread
+
+   Delegate* mDelegate{nullptr};
+   std::unique_ptr<IPCServer> mServer;
+
+   //Variables below is accessed only from worker threads
+
+   details::InputMessageReader mMessageReader;
+
+   //called on main thread only!
+   void StartHost()
+   {
+      auto server = std::make_unique<IPCServer>(*this);
+      if(!PluginHost::Start())
+         mDelegate->OnInternalError("Cannot start plugin host");
+      else
+         mServer = std::move(server);
+   }
+   
+   void HandleInternalError(const wxString& msg) noexcept
+   {
+      try
+      {
+         BasicUI::CallAfter([this, self = weak_from_this(), msg]
+         {
+            if(auto lck = self.lock() && mDelegate != nullptr)
+               mDelegate->OnInternalError(msg);
+         });
+      }
+      catch(...)
+      {
+         //no way to report about a problem, though shouldn't be the case...
+      }
+   }
+
+   void HandleResult(details::PluginValidationResult&& result) noexcept
+   {
+      try
+      {
+         BasicUI::CallAfter([this, self = weak_from_this(), result = details::PluginValidationResult { result }]
+         {
+            if(auto lck_self = self.lock())
+            {
+               if(mDelegate == nullptr)
+                  return;
+
+               //Release the current value to keep state invariant
+               //Caller is free to use Validate now
+               std::optional<wxString> request;
+               {
+                  std::lock_guard lck_sync(mSync);
+                  mRequest.swap(request);
```

Comment:
If you really want to minimize the time inside the lock_guard:  Note that swap of an optional with another empty wone does perform a move-construction, see https://en.cppreference.com/w/cpp/utility/optional/swap ...

BUT for wxString that performs a string copy anyway, because wxWidgets is an old library not defining move constructors.

You can improve this using `optional<std::string>` or you can use the member function `wxString::swap` on *mRequest.  (Then `mRequest.has_value()` is still true, it has an empty string in it; so `mRequest.reset()` would be needed.)

Reply:
You also used `optional<wxString>` in PluginHost and might change it.

Reply:
You didn't take this suggestion yet in the lastest force-push.  Did you not understand it?